rfcs/2023-05-03-data-volume-metrics.md

```diff
@@ -0,0 +1,174 @@
+  - `component_sent_event_total`
+- The metrics sent by each sink needs to be tagged with the source id of the
+  event so the route an event takes through Vector can be queried.
+- Each event needs to be labelled with a `service`. This is a new concept
+  within Vector and represents the application that generated the log,
+  metric or trace.
+- The service tag and source tag in the metrics needs to be opt in so customers
+  that don't need the increased cardinality are unaffected.
+
+### Out of scope
+
+- Separate metrics, `component_sent_bytes_total`  and `component_received_bytes_total`
+  that indicate network bytes sent by Vector are not considered here.
+
+## Pain
+
+Currently it is difficult to accurately gauge the volume of data that is moving
+through Vector. It is difficult to query where data being sent out has come
+from.
+
+## Proposal
+
+### User Experience
+
+Global config options will be provided allowing the name of the `service` tag and the
+`source` tag to be specified. For example:
+
+```yaml
+telemetry:
+  tags:
+    service: theservice
+    source: theinput
+```
+
+This will cause Vector to emit a metric like (note the last two tags):
+
+```statds
+vector.component_sent_event_bytes_total:123|c|#component_id:out,component_kind:sink,component_name:out,component_type:console,host:machine,theservice:somekindofservice,theinput:stdin
+```
+
+The default will be to not emit these tags.
+
+### Implementation
+
+#### Metric tags
+
+**service** - to attach the service, we need to add a new meaning to Vector - 
+              `service`. Any sources that receive data that could potentially 
+              be considered a service will need to indicate which field means 
+              `service`. This work has largely already been done with the 
+              LogNamespacing work, so it will be trivial to add this new field.
+
+**source** - A new field will be added to the [Event metadata][event_metadata] - 
+             `Arc<OutputId>` that will indicate the source of the event. 
+             `OutputId` will need to be serializable so it can be stored in the
+             disk buffer. If the event is loaded from the buffer and it is pointing
+             to a source that no longer exists, given that it is just an identifier
+             it won't cause any issues.
+
+We will need to do an audit of all components to ensure the
+bytes emitted for the `component_received_event_bytes_total` and
+`component_sent_event_bytes_total` metrics are the estimated JSON size of the
+event.
+
+These tags will be given the name that was configured in [User Experience]
+(#user-experience).
+
+Transforms `reduce` and `aggregate` combine multiple events together. In this
+case the `source` and `service` of the first event will be taken.
+
+If there is no `source` specified (the event was created by the `lua` transform)
+- a source of `-` will be emitted.
+
+If there is no `service` available, a service of `-` will be emitted.
```

Comment:
I'd like to see a rationale for mandating these stub tag values as opposed to just eliding the tags when not present.

Reply:
Agreed, we should call that out explicitly here, but we wouldn't want to publish without the tag since that would indicate that the metric represents the aggregate value across all `service`s.